In [1]:
import pandas as pd
import numpy as np

In [2]:
columns = ['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket', 'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest']
df = pd.read_csv('Zbiór danych Titanic.csv', sep=',',skiprows = 17, names=columns, header=None )
df.head(20)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex     age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female      29      0      0     24160  211.3375       B5        S    2   
1     male  0.9167      1      2    113781    151.55  C22 C26        S   11   
2   female       2      1      2    113781    151.55  C22 C26        S    ?   
3     male      30      1      2    113781    151.55  C22 C26        S    ?   
4   female      25      1      2    113781    151.55  C22 C26        S    ?   
5     male      48      0      0     19952     26.55      E12        S    3   
6   female      63      1      0     13502   77.9583       D7        S   10   
7     male      39      0      0    112050         0      A36        S    ?   
8   female      53      2      0     11769   51.4792     C101        S    D   
9     male      71      0      0  PC 17609   49.5042        ?        C    ?   
10    male      47      1      0  PC 17757   227.525  C62 C64        C    ?   
11  female      18      1      0  PC 17757   227.525  C62 C64        C    4   
12  female      24      0      0  PC 17477      69.3      B35        C    9   
13  female      26      0      0     19877     78.85        ?        S    6   
14    male      80      0      0     27042        30      A23        S    B   
15    male       ?      0      0  PC 17318    25.925        ?        S    ?   
16    male      24      0      1  PC 17558  247.5208  B58 B60        C    ?   
17  female      50      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female      32      0      0     11813   76.2917      D15        C    8   
19    male      36      0      0     13050   75.2417       C6        C    A   

   body                        home.dest  
0     ?                     St Louis, MO  
1     ?  Montreal, PQ / Chesterville, ON  
2     ?  Montreal, PQ / Chesterville, ON  
3   135  Montreal, PQ / Chesterville, ON  
4     ?  Montreal, PQ / Chesterville, ON  
5     ?                     New York, NY  
6     ?                       Hudson, NY  
7     ?                      Belfast, NI  
8     ?              Bayside, Queens, NY  
9    22              Montevideo, Uruguay  
10  124                     New York, NY  
11    ?                     New York, NY  
12    ?                    Paris, France  
13    ?                                ?  
14    ?                    Hessle, Yorks  
15    ?                     New York, NY  
16    ?                     Montreal,

# Zadanie 1

In [3]:
for column in df.columns:
  print(f'Liczba etykiet zmiennej {column}:{len(df[column].unique())}')

Liczba etykiet zmiennej pclass:3
Liczba etykiet zmiennej survived:2
Liczba etykiet zmiennej name:1307
Liczba etykiet zmiennej sex:2
Liczba etykiet zmiennej age:99
Liczba etykiet zmiennej sibsp:7
Liczba etykiet zmiennej parch:8
Liczba etykiet zmiennej ticket:929
Liczba etykiet zmiennej fare:282
Liczba etykiet zmiennej cabin:187
Liczba etykiet zmiennej embarked:4
Liczba etykiet zmiennej boat:28
Liczba etykiet zmiennej body:122
Liczba etykiet zmiennej home.dest:370


In [4]:
most_common_names = df['name'].value_counts().nlargest(2).index
df[df['name'].isin(most_common_names)]

pclass  survived                  name     sex   age  sibsp  parch  \
725       3         1  Connolly, Miss. Kate  female    22      0      0   
726       3         0  Connolly, Miss. Kate  female    30      0      0   
924       3         0      Kelly, Mr. James    male  34.5      0      0   
925       3         0      Kelly, Mr. James    male    44      0      0   

     ticket    fare cabin embarked boat body home.dest  
725  370373    7.75     ?        Q   13    ?   Ireland  
726  330972  7.6292     ?        Q    ?    ?   Ireland  
924  330911  7.8292     ?        Q    ?   70         ?  
925  363592    8.05     ?        S    ?    ?         ?

# Zadanie 2

In [5]:
print(f'Liczba wszystkich pasażerów to {len(df)}')

Liczba wszystkich pasażerów to 1309


# Zadanie 3

Tak jak zostało już wspomnanie w poprzednim notatnku mamy 1309 obserwacji i 14 kolumn. Sprawdziłam czy wiersze nie są zduplikowane ponieważ wartości w kolumnie name nie były unikalne. Wydaje się, że nie jest to bład i dwie osoby dwukrtonie nazywają się tak samo/mogła nastąpić pomyłka osoby, która zbierała te dane. Ten fakt nie jest jednak aż tak znaczący do dalszych analiz.
Zmienne o małej kardynalności to:
- pclass
- survied
- sex
- sibsp
- parch
- embarked
- boat ( w tym przypadku powiedziałabym średnia)

Zmienne o dużej kardynalności:
- name
- age
- ticket
- fare
- cabin
- body
- home.dest

# Zadanie 4

In [6]:
uniq = np.unique(df['cabin'], return_counts = True )[1]
unique_count = np.array([len(uniq)])
print(f'Istnieje {unique_count[0]} unikalnych wartości dla zmiennej cabin' )

Istnieje 187 unikalnych wartości dla zmiennej cabin


# Zadanie 5

In [7]:
df['CabinReduced'] = df['cabin'].astype(str).str[0]

In [8]:
df[['cabin', 'CabinReduced']].head(20)

cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9         ?            ?
10  C62 C64            C
11  C62 C64            C
12      B35            B
13        ?            ?
14      A23            A
15        ?            ?
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

# Zadanie 6

In [9]:
print(f'Liczba etykiet zmiennej  CabinReduced: {len(df["CabinReduced"].unique())}')

Liczba etykiet zmiennej  CabinReduced: 9


In [10]:
reduction =  ( len(df["cabin"].unique()) - len(df["CabinReduced"].unique()) ) / len(df["cabin"].unique()) * 100
print(f'Kardynalność została zredukowana o {reduction:.2f}%')

Kardynalność została zredukowana o 95.19%


# Zadanie 7

Decydując się na wycieciu tylko literki z numery kabiny redukujemy kardnalność naszego zbioru zyskująć dodatkowo bardziej uniwersalną i łatwą w interpretacji informację. Bez problemu możemy porównywać na jakim poziomie znajdywała się czyjaś kabina. Tracimy jednak możliwość brania pod uwagę w naszych dalszych analizach rozmieszczenie konkrentych pokoi i szukania zależnośći przykładowo pomiędzy tym czy osoby o mniejszych numerach częściej przeżywały. Jednak biorąc pod uwagę fakt jak bardzo kardynalność tej kolumny została pomniejszona strata ta nie jest aż tak znacząca.